### DISCUSSION TITLE
Unable to install bugsnag-cli in post-publish script

### DISCUSSION THREAD

**Discussion by jeroenbell** [codemagic-ci-cd/codemagic-docs#2023] [Q&A] [answered]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Flutter (Dart)

### Steps to reproduce

1. Add a post-publish script:
```
#!/bin/sh
set -e

BUGSNAG_CLI_PATH=$HOME/.local/bugsnag/bin/bugsnag-cli

echo '--- Installing bugsnag-cli for:' $(uname)-$(uname -m) '---'
curl -o- https://raw.githubusercontent.com/bugsnag/bugsnag-cli/main/install.sh | bash

echo '--- Uploading Android mappings ---'
$BUGSNAG_CLI_PATH upload android-aab $CM_BUILD_OUTPUT_DIR/app-release.aab --api-key=$BUGSNAG_API_KEY
```
2. Start a build.

### Expected results

A successful install, which happens _sometimes_:
```
--- Installing bugsnag-cli for: Darwin-arm64 ---
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  4029  100  4029    0     0  35187      0 --:--:-- --:--:-- --:--:-- 36627
==> This script will install:
/Users/builder/.local/bugsnag/bin/bugsnag-cli
==> The following new directories will be created:
/Users/builder/.local/bugsnag/bin
==> Downloading and installing Bugsnag CLI...
#=#=#                                                                         
##O#-#                                                                        

                                                                           0.2%
                                                                           1.3%
########################################                                  55.8%
##########################################################                80.8%
######################################################################## 100.0%
Warning: /Users/builder/.local/bugsnag/bin is not in your PATH.
  Instructions on how to configure your shell for Bugsnag CLI
  can be found in the 'Next steps' section below.
==> Installation successful!
```

### Actual results

More often than not, the curl fails:
```
--- Installing bugsnag-cli for: Darwin-arm64 ---
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  4029  100  4029    0     0  28944      0 --:--:-- --:--:-- --:--:-- 30067
==> This script will install:
/Users/builder/.local/bugsnag/bin/bugsnag-cli
==> The following new directories will be created:
/Users/builder/.local/bugsnag/bin
==> Downloading and installing Bugsnag CLI...
Failed to get download URL from https://api.github.com/repos/bugsnag/bugsnag-cli/releases/latest


Build failed :|
Step 9 script `Post-publish script` exited with status code 1
```
If you inspect the [script](https://raw.githubusercontent.com/bugsnag/bugsnag-cli/main/install.sh), it will try to find the `browser_download_url` (from https://api.github.com/repos/bugsnag/bugsnag-cli/releases/latest), for `arm64-macos-bugsnag-cli` (using the `$(uname)` and `$(uname -m)` variables) which is a valid entry. I'm left guessing as to why it fails most of the times 🤷 .

### Build id (optional)

64f884abcbc8012dc5a6a4c3

**Comment by rafeysup93** [comment]
Hi @jeroenbell,

We have been able to reproduce the same on our end. We are investigating this issue.

**Reply by rafeysup93** [reply]
I have investigated this issue, and it is an intermittent issue that I reproduced on my local machine. I suggest you raise this issue to BugSnag, as this is not limited to our Virtual machines.

**Reply by Frank3K** [reply]
Thanks for investigating. We've reported the issue to bugsnag: https://github.com/bugsnag/bugsnag-cli/issues/57.

**Comment by Frank3K** [comment]
**Update**

We've further investigated the issue and we've found that it fails when the rate limit is reached. For unauthenticated users it's 60 request per hour ([source](https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#rate-limits-for-requests-from-personal-accounts)). On a local machine it's not that easy to reach, but on CI platforms like codemagic it is since the limit is per IP.

With some more verbose logging:

```sh
> GET /repos/bugsnag/bugsnag-cli/releases/latest HTTP/2
> Host: api.github.com
> User-Agent: curl/8.1.2
> Accept: */*
>
< HTTP/2 403
< date: Fri, 22 Sep 2023 06:51:48 GMT
< server: Varnish
< strict-transport-security: max-age=31536000; includeSubdomains; preload
< x-content-type-options: nosniff
< x-frame-options: deny
< x-xss-protection: 1; mode=block
< content-security-policy: default-src 'none'; style-src 'unsafe-inline'
< access-control-allow-origin: *
< access-control-expose-headers: ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-RateLimit-Used, X-RateLimit-Resource, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, Deprecation, Sunset
< content-type: application/json; charset=utf-8
< referrer-policy: origin-when-cross-origin, strict-origin-when-cross-origin
< x-github-media-type: github.v3; format=json
< x-ratelimit-limit: 60
< x-ratelimit-remaining: 0
< x-ratelimit-reset: **********
< x-ratelimit-resource: core
< x-ratelimit-used: 60
< content-length: 281
< x-github-request-id: ****:*****:********:*******:*********
<
{"message":"API rate limit exceeded for ***.***.***.***. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)","documentation_url":"https://docs.github.com/rest/overview/resources-in-the-rest-api#rate-limiting"}
* Connection #0 to host api.github.com left intact
Failed to get download URL from https://api.github.com/repos/bugsnag/bugsnag-cli/releases/latest
```

So definetely not a codemagic issue, but something bugsnag can fix.

**Comment by Frank3K** [comment, marked as answer]
**Update**

The bugsnag team has a released a fix for this issue: https://github.com/bugsnag/bugsnag-cli/issues/57#issuecomment-1770696417.
